[DWARFLinker][DWARFLinkerParallel][NFC] Refactor DWARFLinker&DWARFLinkerParallel to have a common library. Part 1.

avl-llvm opened the pull request:
This patch is extracted from https://github.com/llvm/llvm-project/pull/74725 .
It creates DWARFLinkerBase library, places DWARFLinker code into DWARFLinker\Apple directory, puts code from DWARFLinker\Apple into 'dwarflinker' namespace, places DWARFLinkerParallel into DWARFLinker\LLVM directory, updates BOLT to use new library. This patch is NFC.

avl-llvm committed:
[DWARFLinker][DWARFLinkerParallel][NFC] Refactor DWARFLinker&DWARFLinkerParallel to have a common library. Part 1.

This patch creates DWARFLinkerBase library, places DWARFLinker code into
DWARFLinker\Apple, places DWARFLinkerParallel into DWARFLinker\LLVM.
This patch is NFC.

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+    Object,
+    Assembly,
+  };
+  /// The kind of accelerator tables we should emit.
```
Since we have a chance to rewrite some of these comments, let's make them more impersonal and specific:

"The kind of accelerator tables to be emitted."

felipepiovezan replied:
I think this comment has not addressed

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+    Pub,       ///< .debug_pubnames, .debug_pubtypes
+    DebugNames ///< .debug_names.
+  };
+  /// Add object file to be linked. Pre-load compile unit die. Call
```
> Add object file to be linked

Add *an* object file to be linked.

> Pre-load compile unit die

Can you elaborate on this sentence? I'm not sure what pre-loading means.

>  If specified \p File has reference to the Clang module then such module would be pre-loaded by \p Loader for !Update case.

I'm not sure I understand this, but if I am reading this correctly, we can say something like:

"If \p File has reference to a Clang module then the module is be pre-loaded by \p Loader ... ?"

Actually, what is the meaning of "pre-loaded by \p Loader for !Update case."?

avl-llvm replied:
> > Add object file to be linked
> 
> Add _an_ object file to be linked.
> 
> > Pre-load compile unit die
> 
> Can you elaborate on this sentence? I'm not sure what pre-loading means.

When an object file is added to be linked, part of input DWARF Context corresponding to this object file is loaded. Specifically, The DWARFUnit::getUnitDIE( /* ExtractUnitDIEOnly == true */) is called for each compile unit:

```
    for (const std::unique_ptr<DWARFUnit> &CU :
         ObjectContexts.back().File.Dwarf->compile_units()) {
      DWARFDie CUDie = CU->getUnitDIE();
   }

```
> 
> > If specified \p File has reference to the Clang module then such module would be pre-loaded by \p Loader for !Update case.
> 
> I'm not sure I understand this, but if I am reading this correctly, we can say something like:
> 
> "If \p File has reference to a Clang module then the module is be pre-loaded by \p Loader ... ?"
> 
> Actually, what is the meaning of "pre-loaded by \p Loader for !Update case."?

Clang module is represented as a compile unit referencing another file. If such Clang module reference is discovered during adding object file then DWARFLinker goes through the reference, load DWARFFile by calling ErrorOr<DWARFFile &> ObjFileLoaderTy(StringRef ContainerName, StringRef Path), creates CompileUnit for referenced file, adds created CompileUnit into the "ModuleUnits" list. Expansion of clang module reference is done only in !Update case.

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  /// has reference to the Clang module then such module would be
+  /// pre-loaded by \p Loader for !Update case.
+  ///
+  /// \pre NoODR, Update options should be set before call to addObjectFile.
```
Is this equivalent to:

"\pre a call to `setNoODR(true)` must be made" ?

avl-llvm replied:
Not exactly, probably this would be an equivalent:

"\pre a call to setNoODR(true) and/or setUpdateIndexTablesOnly(bool Update) must be made when required."

Is it OK?

felipepiovezan replied:
Sounds good!

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  virtual void addObjectFile(
+      DWARFFile &File, ObjFileLoaderTy Loader = nullptr,
+      CompileUnitHandlerTy OnCUDieLoaded = [](const DWARFUnit &) {}) = 0;
+  /// Link debug info for added objFiles. Object files are linked all together.
```
>/// Link debug info for added objFiles.

Link the debug info for all object files added through calls to `addObjectFile`?

> Object files are linked all together

This sentence makes it sound like we are linking object files, but we are only linking debug info? IIUC we should remove this entirely

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  virtual Error link() = 0;
+  /// A number of methods setting various linking options:
+  /// Allows to generate log of linking process to the standard output.
+  virtual void setVerbosity(bool Verbose) = 0;
```
"Enable logging to standard output"

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  virtual void setVerifyInputDWARF(bool Verify) = 0;
+  /// Do not unique types according to ODR.
+  virtual void setNoODR(bool NoODR) = 0;
+  /// Update index tables only(do not modify rest of DWARF).
```
space before "("

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  virtual void setNoODR(bool NoODR) = 0;
+  /// Update index tables only(do not modify rest of DWARF).
+  virtual void setUpdateIndexTablesOnly(bool Update) = 0;
+  /// Allow generating valid, but non-deterministic output.
```
I don't think we need to say "valid" here, this is redundant. I'd focus on explaining what can be gained by allowing this.

"Allows generating non-deterministic output in exchange for more parallelism"?

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  /// Allow generating valid, but non-deterministic output.
+  virtual void setAllowNonDeterministicOutput(bool) = 0;
+  /// Set whether to keep the enclosing function for a static variable.
+  virtual void setKeepFunctionForStatic(bool KeepFunctionForStatic) = 0;
```
Is this meant to be "to keep the DIE of a function where a function-scope static variable is declared"?

Do we usually drop these DIEs? Might be something worth documenting and explaining.

avl-llvm replied:
yes we usually drop these DIEs. It has some description here:

https://llvm.org/docs/CommandGuide/dsymutil.html#cmdoption-dsymutil-keep-function-for-static

Do you suggest to extend that description?

felipepiovezan replied:
Personally, I would have been a bit more precise on describing what's being kept. But with the context you provided, it's possible to figure out what this is

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  /// Set whether to keep the enclosing function for a static variable.
+  virtual void setKeepFunctionForStatic(bool KeepFunctionForStatic) = 0;
+  /// Use specified number of threads for parallel files linking.
+  virtual void setNumThreads(unsigned NumThreads) = 0;
```
Does this method make sense for all linkers?

avl-llvm replied:
yes, it does. Apple linker is able to work parallelly utilizing up to 2 threads.  So this method can switch between multi-thread and single-thread execution.

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  /// Use specified number of threads for parallel files linking.
+  virtual void setNumThreads(unsigned NumThreads) = 0;
+  /// Add kind of accelerator tables to be generated.
+  virtual void addAccelTableKind(AccelTableKind Kind) = 0;
```
Do _all_ linkers allow multiple calls to this? Or should this be more of a "set" instead of an "add"?

avl-llvm replied:
At present moment no linkers have use cases when tables can be added. But from the general point of view we can want to generate different sets of accel tables and then it is good if interface allows this.

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  /// Add kind of accelerator tables to be generated.
+  virtual void addAccelTableKind(AccelTableKind Kind) = 0;
+  /// Set prepend path for clang modules.
+  virtual void setPrependPath(const std::string &Ppath) = 0;
```
StringRef?

felipepiovezan replied:
I noticed you change this, which is good, but you also changed the _member variable_ from `std::string` to `StringRef`. This may be fine, but you need to make sure the variable lifetimes are ok. My original comment was just about the interface, as I am not familiar with the lifetime of the variables here to be certain it is safe to _store_ a StringRef

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -0,0 +1,98 @@
+  /// Set estimated objects files amount, for preliminary data allocation.
+  virtual void setEstimatedObjfilesAmount(unsigned ObjFilesNum) = 0;
+  /// Set verification handler which would be used to report verification
+  /// errors.
```
which would be used -> used

felipepiovezan reviewed (commented):
Since we now have a base class for linker implementations, I looked at its documentation from the perspective of "what do I need to do in order to implement a new linker".

Left some wording suggestions! Hopefully others will have a look as well

felipepiovezan commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -79,11 +82,10 @@ class DWARFLinkerBase {
   virtual void setEstimatedObjfilesAmount(unsigned ObjFilesNum) = 0;
-  /// Set verification handler which would be used to report verification
-  /// errors.
+  /// Set verification handler which used to report verification errors.
```
~~which~~ used

avl-llvm commented on llvm/include/llvm/DWARFLinker/DWARFLinkerBase.h:
```diff
@@ -42,35 +42,38 @@ class DWARFLinkerBase {
     Assembly,
   };
-  /// The kind of accelerator tables we should emit.
+  /// The kind of accelerator tables to be emitted.
```
The original comment is addressed here.

avl-llvm commented:
friendly ping

felipepiovezan reviewed (approved):
This LGTM, but it would be nice to have at least one more person looking into it.
@JDevlieghere ?

JDevlieghere commented:
The base class looks good but I would prefer a slight change to the directory layout and namespaces. Particularly the distinction between "Apple" and "LLVM" bother me. I understand how you ended up there, but I think it unnecessary associates the algorithm with particular vendors. How about "Parallel" and "Classic"? 

With that naming scheme, what I'd like this to look like would be:  

- The `DWARFLinker/` root directory contains shared classes and lives in the `dwarflinker` namespace. I think it should probably be `dwarf_linker` but I'm fine with `dwarflinker` if that's too much churn. 
- The `DWARFLinker/Classic` directory containing classes needed for the dsymutil lockstep algorithm. Everything here should live under the `dwarf(_)linker::classic` namespace. 
- The `DWARFLinker/Parallel` directory containing classes needed for the parallel algorithm. Everything here should live under the `dwarf(_)linker::parallel` namespace.

avl-llvm commented:
> The base class looks good but I would prefer a slight change to the directory layout and namespaces. Particularly the distinction between "Apple" and "LLVM" bothers me. I understand how you ended up there, but I think it unnecessary associates the algorithm with particular vendors. How about "Parallel" and "Classic"?

That is perfectly fine for me. Would you like to change command line option also? 

dsymutil --linker=[classic,parallel] 

Or it would be only directories/namespaces change only?

> 
> With that naming scheme, what I'd like this to look like would be:
> 
>     * The `DWARFLinker/` root directory contains shared classes and lives in the `dwarflinker` namespace. I think it should probably be `dwarf_linker` but I'm fine with `dwarflinker` if that's too much churn.
> 
>     * The `DWARFLinker/Classic` directory containing classes needed for the dsymutil lockstep algorithm. Everything here should live under the `dwarf(_)linker::classic` namespace.
> 
>     * The `DWARFLinker/Parallel` directory containing classes needed for the parallel algorithm. Everything here should live under the `dwarf(_)linker::parallel` namespace.

JDevlieghere commented:
> That is perfectly fine for me. Would you like to change command line option also?
> 
> dsymutil --linker=[classic,parallel]
> 
> Or it would be only directories/namespaces change only?

I don't feel strongly about the command line options. We're not using them in production yet so the only churn would be the tests. I'll let you decide whatever you think is best.

avl-llvm commented:
> > That is perfectly fine for me. Would you like to change command line option also?
> > dsymutil --linker=[classic,parallel]
> > Or it would be only directories/namespaces change only?
> 
> I don't feel strongly about the command line options. We're not using them in production yet so the only churn would be the tests. I'll let you decide whatever you think is best.

I think it would be good to rename options also. Now is the good moment for this(as options are not used widely). Though, I am going to do this in a separate patch.

ayermolo commented on llvm/lib/DWARFLinker/LLVM/DependencyTracker.cpp:
```diff
@@ -9,8 +9,9 @@
 
-namespace llvm {
-namespace dwarflinker_parallel {
+using namespace llvm;
```
why switch to using here?

avl-llvm replied:
This style is recommended by guidelines for namespaces in source file : https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions.
Here is the example of using it in codebase - https://github.com/llvm/llvm-project/blob/780a5116ba68ec8c53b65008b3407479478b2d5e/llvm/lib/DebugInfo/GSYM/DwarfTransformer.cpp#L26

The patch changes namespace structure so we need to update this place anyway. So it is updated using the recommended form.

avl-llvm commented:
addressed comments(renamed directories Apple->Classic, LLVM->Parallel), renamed "dwarflinker" namespace into "dwarf_linker", renamed "dwarflinker_parallel" namespace into "parallel", added "classic" namespace as a child of "dwarf_linker").

JDevlieghere reviewed (approved):
Thanks @avl-llvm, LGTM!

avl-llvm committed:
remove redundand option processing.

avl-llvm commented:
> The library names don't match the directory layout, which is really unusual.
> 
> How about naming the library in DWARFLinker/Classic LLVMDWARFLinkerClassic to match the name? Also, the one in DWARFLinker should then either be renamed to LLVMDWARFLinker or be moved into DWARFLinker/Base.

yep, will rename libraries:

DWARFLinker->LLVMDWARFLinker
DWARFLinker/Classic -> LLVMDWARFLinkerClassic

Thanks!
